src/Agent.Worker/TaskRunner.cs

```diff
@@ -25,27 +26,106 @@ public sealed class TaskRunner : AgentService, ITaskRunner
 
-        public Task<TaskResult> RunAsync()
+        public async Task RunAsync()
         {
+            // Validate args.
             Trace.Entering();
             ArgUtil.NotNull(ExecutionContext, nameof(ExecutionContext));
+            ArgUtil.NotNull(ExecutionContext.Variables, nameof(ExecutionContext.Variables));
             ArgUtil.NotNull(TaskInstance, nameof(TaskInstance));
+            var definitionManager = HostContext.GetService<ITaskDefinitionManager>();
+            var handlerFactory = HostContext.GetService<IHandlerFactory>();
 
-            ExecutionContext.Start();
-            CancellationTokenSource ts = new CancellationTokenSource(TimeSpan.FromMinutes(1));
-            while (!ts.IsCancellationRequested)
+            // Load the task definition and choose the handler.
+            Definition definition = definitionManager.Load(TaskInstance);
+            ArgUtil.NotNull(definition, nameof(definition));
+            HandlerData handlerData =
+                definition.Data?.Execution?.All
+                .OrderBy(x => !x.MatchesPlatform()) // Sort true to false.
+                .ThenBy(x => x.Priority)
+                .FirstOrDefault();
+            if (handlerData == null)
             {
-                ExecutionContext.Write(null, DateTime.UtcNow.ToString("O"));
-                Task.Delay(50).Wait();
+                // TODO: BETTER ERROR AND LOC
+                throw new Exception("Supported handler not found.");
             }
-            ExecutionContext.Result = TaskResult.Succeeded;
-            ExecutionContext.Complete();
 
-            // TODO: Load the task definition Path.Combine("tasks", TaskInstance.Name, taskInstance.Version)
-            // TODO: Choose the handler.
-            return Task.FromResult(TaskResult.Succeeded);
+            // Load the default input values from the definition.
+            Trace.Verbose("Loading default inputs.");
+            var inputs = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var input in (definition.Data?.Inputs ?? new TaskInputDefinition[0]))
+            {
+                string key = input?.Name?.Trim() ?? string.Empty;
+                if (!string.IsNullOrEmpty(key))
+                {
+                    inputs[key] = input.DefaultValue?.Trim() ?? string.Empty;
+                }
+            }
+
+            // Merge the instance inputs.
+            Trace.Verbose("Loading instance inputs.");
+            foreach (var input in (TaskInstance.Inputs as IEnumerable<KeyValuePair<string, string>> ?? new KeyValuePair<string, string>[0]))
+            {
+                string key = input.Key?.Trim() ?? string.Empty;
+                if (!string.IsNullOrEmpty(key))
+                {
+                    inputs[key] = input.Value?.Trim() ?? string.Empty;
+                }
+            }
+
+            // Expand the inputs.
+            Trace.Verbose("Expanding inputs.");
```

Comment:
Again there might be a better home for this logic after the job variable expansion logic is added. The macro replacement logic here is different from the logic in the handler data class, so it didn't make sense to put it into a common place at this point.